Fixes #4419 - rearranging smart class parameters edit form

orrabin opened the pull request:
opening this so it can get reviewed by more people

domcleal commented:
Are all of the commits meant to be under http://projects.theforeman.org/issues/10467, or what other issues do they address precisely?

unorthodoxgeek commented:
The commit which references 10467 fixes 10467
this PR is for the 4470 tracker as a whole

edit: I'm completely incorrect in this comment, please ignore me, and ask @orrabin not to shoot me :smile:

domcleal commented:
> this PR is for the 4470 tracker as a whole

Er, not all of it, surely?  It seems to be me that the scope of this PR, at least at the moment, is somewhat smaller.

unorthodoxgeek commented:
hence the WIP flag

domcleal commented:
This strikes me as a particularly bad idea, incremental fixes for specific issues would be a much easier way to work.  I'm happy to see parts of our parameter support be rewritten, but don't want to see one mega-PR where we have no ability to review it due to the size.

orrabin commented:
@domcleal this is just the first step of UI changes discussed here: https://groups.google.com/forum/#!searchin/foreman-dev/parameters$20usability/foreman-dev/qUFogNWbOe0/dRbognPrlG0J
I'll open up an issue for them.
We will open up individual PRs for each issue if we can (in this case some of my changes where used by @unorthodoxgeek so we opened one PR)

ohadlevy commented:
can you please add a few screenshots for the changes?

thanks :)

domcleal commented:
> We will open up individual PRs for each issue if we can

So close this and open up more specific PRs?

orrabin commented:
@domcleal there was an issue open: http://projects.theforeman.org/issues/4419
these 2 issues are related and should be fixed together

domcleal commented:
Okay, let's rename this PR and commits then to fixes #4419, if that's what it's specifically going to solve.  I'd just like to understand, and put a hard limit on the scope before it gets crazy.

domcleal commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,62 @@
+                            :placeholder => _("Value")) %>
+              <span class="hidden_value">
+                <%= check_box_tag "host[lookup_values_attributes][#{key.id}][hidden]" %>
+                Hidden
```
i18n extraction

domcleal commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,62 @@
+
+              <span class="puppet-default">
+                <%= check_box_tag "host[lookup_values_attributes][#{key.id}][use_puppet_default]", nil, value.try(:use_puppet_default) %>
+                Use Puppet Default
```
extract, lower case 'D'

domcleal commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,62 @@
+              </span>
+          </td>
+          <td>
+            <%= link_to(_("edit"), nil, :title => _("Edit this value"),
```
what's the purpose of this, as there's a field to the left already?

unorthodoxgeek replied:
it serves no purpose, this is what happens when I follow the wireframe without thinking about it too much :facepalm:

domcleal commented on app/views/lookup_keys/_values.html.erb:
```diff
@@ -0,0 +1,23 @@
+    <th colspan='2' class='col-md-5'><%= _('Attribute Type') %> <span class="help-inline"><%= popover("", _("Matcher is a combination of an attribute and its value, if they match, the value below would be provided.<br> You may use any attribute foreman knows about, such as facts etc for example: <code> domain = example.com </code> or <code> is_virtual = true</code>"),
+                                                                :title => _("Explain matchers")).html_safe %></span></th>
+    <th class='col-md-4'><%= _('Value') %></th>
+    <th class='col-md-3'><%= _('Use Puppet Default') %><span class="help-inline"> <%= use_puppet_default_help() %></span></th>
```
lower case 'Default'

domcleal replied:
This is actually a very wide heading too, it takes up a lot of horizontal room on the form when you consider it's just a checkbox:

![screenshot from 2015-05-22 12 43 03](https://cloud.githubusercontent.com/assets/482089/7769879/294a5322-0080-11e5-8ca1-c9775e6eef30.png)

orrabin replied:
When I tried to change it to `col-md-2` it became a two line header.
Maybe if I change it to 'Puppet default' that would be enough?

domcleal replied:
Yes, that'd help.

domcleal commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,62 @@
+        <tr class="lookup_key_row <%="overridden" if key.overridden?(obj) %>" data-lookup-key-id="<%=key.id%>">
+          <td class='col-md-2'>
+            <%= content_tag :span, key.key, :class => "col-md-2 lookup_key_name", :'data-property' => 'name' %></td>
+          <td class='col-md-7'><%= obj.class.model_name == "Host" ? host_key_with_diagnostic(obj, value_hash, key) : hostgroup_key_with_diagnostic(obj, key) %></td>
```
It looks rather odd for the diagnostics to have a strikethrough when it contains parameter-level info (data type, description), not just info about where the value came from.

domcleal replied:
![screenshot from 2015-05-22 12 32 33](https://cloud.githubusercontent.com/assets/482089/7769739/af407670-007e-11e5-95a8-a06691d4722b.png)

domcleal replied:
Secondly, perhaps we could lose the phrase "Additional info" and just put the "i" symbol next to the parameter name, giving us more room for the value text box.

domcleal commented on app/views/common_parameters/_parameters_table.html.erb:
```diff
@@ -0,0 +1,29 @@
+<div class="tab-pane" id="params">
+  <div class="tab-content">
+    <ul class="nav nav-tabs" data-tabs="tabs">
```
I don't think that an additional level of navigation is helpful here, we now have..
1. application level menus
2. host edit form tabs (Host/Puppet Classes/Parameters etc)
3. Puppet Parameters/Global Parameters
4. list of classes

domcleal replied:
Perhaps:
1. make global parameters the first tab in the class list under Puppet Parameters (e.g. "Global", "apache", "ntp")
2. have them as separate host level tabs: Host /  Puppet Classes / Puppet Parameters / Global Parameters / Additional Info
3. roll Puppet class parameters into the Puppet Classes tab

unorthodoxgeek replied:
we opted for a less confusing UI and the puppetclass/global tabs would have a different style so it won't look the same as the host edit tabs.

domcleal commented on app/assets/javascripts/host_edit.js:
```diff
@@ -2,6 +2,56 @@ $(document).on('ContentLoad', function(){onHostEditLoad()});
+    hiddenToggle.change(toggleHidden);
+  }
+
+  override = function() {
```
This doesn't seem to add a strikethrough (override-param class) to the textareas, ditto in the removeOverride function below which doesn't seem to remove it.

domcleal commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,62 @@
+          <td>
+            <%= link_to(_("edit"), nil, :title => _("Edit this value"),
+                                 :'data-tag' => 'edit', :class => "btn btn-default edit_override_button" ) if authorized_via_my_scope("host_editing", "create_params") %>
+            <%= link_to(_("remove"), nil, :title => _("Remove this value"),
```
normally we use an 'x' rather than a button

unorthodoxgeek replied:
it's an removal action, not just hiding it from UI

domcleal replied:
Where do we use 'x' to indicate hiding?

domcleal commented on app/views/lookup_keys/_values.html.erb:
```diff
@@ -0,0 +1,23 @@
+<table class="table table-condensed">
+  <thead>
+  <tr>
+    <th colspan='2' class='col-md-5'><%= _('Attribute Type') %> <span class="help-inline"><%= popover("", _("Matcher is a combination of an attribute and its value, if they match, the value below would be provided.<br> You may use any attribute foreman knows about, such as facts etc for example: <code> domain = example.com </code> or <code> is_virtual = true</code>"),
```
Lower case 'Type', upper case 'Foreman', and this should probably be rewritten a little (remind me later).

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,59 +1,61 @@
+    <fieldset>
+      <legend><%= _("Prioritize Attribute Order") %></legend>
+      <h6><%= popover("", _("The order in which matchers keys are processed, first match wins.<br> You may use multiple attributes as a matcher key, for example, an order of <code>host group, environment</code> would expect a matcher such as <code>hostgroup = \"web servers\", environment = production</code>"),
+                      :title => _("The order in which values are resolved")).html_safe %> <%= _("Set the order in which values are resoved") %></h6>
```
"resolved"

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,59 +1,61 @@
+  <div class="matcher-parent" <%= "id=#{(f.object.key || 'new_lookup_keys').to_s.gsub(' ', '_')}_lookup_key_override_value" %> style=<%= "display:none;" if (f.object.is_param && !f.object.override) %>>
+    <fieldset>
+      <legend><%= _("Prioritize Attribute Order") %></legend>
+      <h6><%= popover("", _("The order in which matchers keys are processed, first match wins.<br> You may use multiple attributes as a matcher key, for example, an order of <code>host group, environment</code> would expect a matcher such as <code>hostgroup = \"web servers\", environment = production</code>"),
```
popover is a different style to the large Documentation button below

orrabin replied:
I don't understand, the Documentation button should look different since it redirects you out of the form

domcleal replied:
Ah yes, never mind.

domcleal replied:
---

(separate thread)

Usually the 'i' is just on fields, not section introductions.  It's also usually after the item, not before.

Perhaps move some of the popover content into the section introduction and remove the popover entirely here?

orrabin replied:
I don't want to make the introduction too long. I could put it on the end of the introduction if that's better.
It should probably be on the `Order` field but it won't look that great right after the list box.

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,59 +1,61 @@
+      <div class="children_fields lookup_values">
+        <%= render 'lookup_keys/values', :f => f, :is_param => is_param %>
+      </div>
+      <%= table_checkbox_f(f, :merge_overrides, :onchange => 'mergeOverridesChanged(this)',
```
The position of these in the form is a bit lost, as they're stuck between a large "Add Matcher-Value" button and a heading, it took me a while to notice them.  It might be better at the top of this section above the table?

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,59 +1,61 @@
+    <%= show_puppet_class f %>
+  </fieldset>
+  <fieldset>
+    <legend><%= _("Add Override") %></legend>
```
This should probably not say "Add" as that suggests a variable number to me (which is actually the matchers further down).  Perhaps "Default value" or "Parameter defaults"?

domcleal commented on app/views/lookup_keys/_values.html.erb:
```diff
@@ -0,0 +1,23 @@
+</tbody>
+</table>
+<%= new_child_fields_template(f, :lookup_values, { :partial => "lookup_keys/value", :form_builder_attrs => { :is_param => is_param, :is_template => true } }) %>
+<%= add_child_link '+ ' + _("Add Matcher-Value"), :lookup_values, { :title => _('add a new matcher-value pair') } %>
```
just "Add matcher" will do I think for consistency

domcleal commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,62 @@
+                            :rows => 1,
+                            :placeholder => _("Value")) %>
+              <span class="hidden_value">
+                <%= check_box_tag "host[lookup_values_attributes][#{key.id}][hidden]" %>
```
Does this do anything yet?  I don't think it's on a lookup value, it might be better as an attribute on the lookup key rather than values.

domcleal commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,62 @@
-<% end %>
+    </thead>
+    <tbody>
+    <% lookup_keys.each_with_index do |key, i| %>
```
this seems to order smart variables after smart class params, should they be sorted alphabetically?  (Actually, I can't see where sorting is done now... probably just the lookup key's default scope.)

domcleal commented on app/views/puppetclasses/_form.html.erb:
```diff
@@ -34,7 +34,7 @@
               <span class="input-group-addon">@</span>
               <%= select_tag "environment_filter", options_from_collection_for_select(@puppetclass.environments, "id", "name"),
-                             :include_blank => "All Environments - (Not filtered)",:class=>'form-control', :onchange=>'filterByEnvironment(this)'%>
+                             :prompt => "All Environments - (Not filtered)",:class=>'form-control', :onchange=>'filterByEnvironment(this)'%>
```
All Envs.. should be extracted

domcleal commented on app/views/puppetclasses/_class_parameters.html.erb:
```diff
@@ -1,18 +1,62 @@
+        <tr id="lookup_value_for_<%=key.id%>" class="lookup_key_override_row <%="visible" if key.overridden?(obj) %>">
+          <% value = key.overriding_value(obj) %>
+          <td class='col-md-2'>
+            <%= content_tag :span, key.key, :class => "col-md-2 lookup_key_name", :'data-property' => 'name' %></td>
```
leave this out perhaps?  It's already on the line above, we could instead have better styling to show the two rows (the original and the override) are one and the same.

e.g. the same background colour, or a little link like we have on the e-mail preferences of users:

![screenshot from 2015-05-22 13 01 26](https://cloud.githubusercontent.com/assets/482089/7770128/c43c06ee-0082-11e5-8740-fd09a2f1991e.png)

unorthodoxgeek replied:
I'll send Kyle an email about this suggestion and let him decide.

unorthodoxgeek replied:
Bottom line of the discussion with Kyle was to keep it as is.

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,59 +1,61 @@
+      <%= textarea_f f, :path, :rows => :auto, :label => _("Order"), :id => 'order', :size => "col-md-8", :onchange => 'fill_in_matchers()', :value => f.object.path %>
+    </fieldset>
+    <fieldset>
+      <legend><%= _("Specify Attribute Matcher") %>  <%= documentation_button('4.2.6SmartMatchers') %></legend>
```
I'll try and come up with some better headings, these are all a bit opaque..

orrabin replied:
thanks, I'll change it once there is a suitable title

orrabin committed:
Fixes #4419 - rearranging smart class parameters edit form

orrabin commented:
@domcleal there are minor changes that are needed in both this and #10467 so we divided the PR, duplicated what was needed and it will be resolved after one is merged and the other is rebased.
We will make sure comments are fixed in both places but there are only 1-2 lines in common so it shouldn't be too bad.

waldenraines commented on app/helpers/layout_helper.rb:
```diff
@@ -158,26 +167,48 @@ def autocomplete_f(f, attr, options = {})
+  def add_label options, f, attr
+    label_size = options.delete(:label_size) || "col-md-2"
+    required_mark = check_required(options, f, attr)
+    label   = options[:label] == :none ? '' : options.delete(:label)
```
Nitpick but why the extra spaces?

orrabin replied:
by mistake, I'll fix

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,61 @@
+    <%= show_puppet_class f %>
+  </fieldset>
+  <fieldset>
+    <legend><%= _("Add Override") %></legend>
```
"Default behavior" or "Default value" perhaps?

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,61 @@
+  </fieldset>
+  <fieldset>
+    <legend><%= _("Add Override") %></legend>
+    <h6><%= _("Override the Puppet class default value") %></h6>
```
I'd suggest expanding this a bit to explain what the section does - no more than one or two sentences.

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,61 @@
+    <legend><%= _("Add Override") %></legend>
+    <h6><%= _("Override the Puppet class default value") %></h6>
+    <%= checkbox_f(f, :override, :onchange => 'toggleOverrideValue(this)', :size => "col-md-8",
+                   :help_inline => popover('', _('Whether the smart variable value is managed by Foreman'))
```
class parameter, not smart variable

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,61 @@
+    <%= textarea_f f, :default_value, :value => f.object.default_value_before_type_cast, :size => "col-md-8",
+                   :disabled => (f.object.is_param && (!f.object.override || f.object.use_puppet_default)),
+                   :fullscreen => :true,
+                   :rows => :auto, :help_inline => popover('', _("Value to use when there is no match")) %>
```
Padding between the text box, the full screen button and the help icon is inconsistent:

![screenshot from 2015-06-25 12 07 01](https://cloud.githubusercontent.com/assets/482089/8353120/c56e23f6-1b32-11e5-9a13-88964ad46f6a.png)

domcleal replied:
Still an issue.

orrabin replied:
do you still see them glued together?
after changing the css I see this:
![image](https://cloud.githubusercontent.com/assets/7894647/8569269/504513b8-257e-11e5-86b2-e2bb836660c1.png)
but maybe I added a margin that was too small?
https://github.com/theforeman/foreman/pull/2398/files#diff-4b492e48ae68ab49ffd1f96675dfa885R26

domcleal replied:
Yes, I still see them next to each other in Firefox.

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,61 @@
+                      :title => _("The order in which values are resolved")).html_safe %> <%= _("Set the order in which values are resolved") %></h6>
+      <%= textarea_f f, :path, :rows => :auto, :label => _("Order"), :id => 'order', :size => "col-md-8", :onchange => 'fill_in_matchers()', :value => f.object.path %>
+    </fieldset>
+    <%= table_checkbox_f(f, :merge_overrides, :onchange => 'mergeOverridesChanged(this)',
```
Should the merge overrides + avoid dupes be on separate lines, like override/puppet default are in the previous section?

orrabin replied:
This was intentional since they are only used in certain types and shouldn't take up so much space.
I could put override/puppet default on the same line but I think they make sense before and after the default value.

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,61 @@
+                         :help_inline => popover("", _("Continue to look for matches after first find (only array/hash type)? Note: merging overrides ignores all matchers that use Puppet default"))) %>
+    <%= table_checkbox_f(f, :avoid_duplicates, :disabled => (!f.object.supports_uniq? || !f.object.merge_overrides), :size => "col-md-1", :label_size => "col-md-4",
+                         :help_inline                    => popover("", _("Avoid duplicate values when merging them (only array type)?"))) %>
+    <fieldset>
```
There's less room between the previous section and this heading than there is between this heading and its content, which looks strange.  Add more padding above?

![screenshot from 2015-06-25 12 10 49](https://cloud.githubusercontent.com/assets/482089/8353190/59e11caa-1b33-11e5-83e1-10fb7019f099.png)

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,61 @@
+    <%= table_checkbox_f(f, :avoid_duplicates, :disabled => (!f.object.supports_uniq? || !f.object.merge_overrides), :size => "col-md-1", :label_size => "col-md-4",
+                         :help_inline                    => popover("", _("Avoid duplicate values when merging them (only array type)?"))) %>
+    <fieldset>
+      <legend><%= _("Specify Attribute Matcher") %>  <%= documentation_button('4.2.6SmartMatchers') %></legend>
```
Perhaps simply "Matchers"?

domcleal commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -1,13 +1,42 @@
+    <tr class="fields matchers form-group <%= 'has-error' if f.object.errors.any? %>">
+      <td>
+        <%= f.hidden_field :match, :class => 'match' %>
+        <%= select_tag '', nil, :class => 'matcher_key' %>
```
When rebased onto develop, this doesn't seem to be activating with select2:

![screenshot from 2015-06-25 12 14 18](https://cloud.githubusercontent.com/assets/482089/8353229/bf079abe-1b33-11e5-989f-7e65f58b020b.png)

orrabin replied:
I'll add it but it's problematic in a table, the only way I can think of adding it is with a max-width in the css and even then you would only see `ho..` selected because the `x` takes up space and there is not enough room for more.
This should be a pretty short list and I don't think it's worth adding it but I'll push the changes and maybe someone will have a suggestion on how to make more space in the matcher row.

domcleal replied:
Probably worth having, since if you have multiple types of overrides, the select2 features are good for filtering and consistency with the rest of the UI.  Maybe move some of the other stuff onto a new line so the value and key boxes are larger.

domcleal commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -1,13 +1,42 @@
+        <%= f.hidden_field :match, :class => 'match' %>
+        <%= select_tag '', nil, :class => 'matcher_key' %>
+        =
+        <%= text_field_tag '', nil, :class => 'matcher_value' %>
```
Should we use autocomplete for this?  It might be a bit of work, so perhaps a separate, future ticket.

orrabin replied:
That is the plan :)

domcleal replied:
Okay, perhaps file it and link to this ticket, if it isn't there already.

orrabin replied:
http://projects.theforeman.org/issues/11049

domcleal commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -1,13 +1,42 @@
+                        :onchange => 'toggleUsePuppetDefaultValue(this, "value")' %>
+      </td>
+      <td>
+          <span class="help-block">
```
inconsistent indentation, and below

domcleal commented on app/views/lookup_keys/_values.html.erb:
```diff
@@ -0,0 +1,24 @@
+<%= add_child_link '+ ' + _("Add Matcher"), :lookup_values, { :title => _('add a new matcher') } %>
+
+
+
```
whitespace

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,61 @@
+      </div>
+    </fieldset>
+    <fieldset>
+      <legend><%= _("Optional input validator") %></legend>
```
Seems odd to move this near the matchers, when it's closer in functionality to the parameter-level data type field.  Why the move?

orrabin replied:
Seems like this is something you change less often than the matchers.
This was an attempt to fit the most important information to one screen without scrolling.

domcleal replied:
In that case wouldn't the matchers be right at the top?  (Not that I'm advocating it.)

If you're changing the data type then you're likely to want to change the validator too, so I don't think they should be split up.  It feels like it should be over multiple tabs, but adding yet more types of tabs here would be ugly and hard to use.

domcleal commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -215,3 +221,60 @@ function validatorTypeSelected(item){
+    var matcher_value = matcher.find('.matcher_value');
+    var order = matcher.closest('.matcher-parent').find('#order').val().split('\n');
+    matcher_key.empty();
+    matcher_key.append("<option></option>");
```
Why the blank?  Is that just for the template ("Add Matcher")?  It seems unnecessary on existing matchers.

domcleal replied:
Still an issue?

orrabin replied:
This was to keep a blank option since I'm emptying `matcher_key` each time I can't use include_blank as it would be deleted anyway

domcleal replied:
Not sure I follow.  If it really can't be removed then you need to prevent the form from being saved when the matcher is set to the blank value - I was able to save one with "blank = Hostgroup".

orrabin replied:
I can remove it but then adding a matcher will by default fill in the key as the first one in the order. I don't mind rejecting blank= or changing this, I'm not sure which is better

domcleal commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -215,3 +221,60 @@ function validatorTypeSelected(item){
+  return match.slice(0, -1);
+}
+
+function fill_in_matchers(){
```
There's some lack of HTML escaping in here, since if you type `<script>alert("hi")</script>` into the Order box, it pops up an alert when building the select menu.

domcleal commented on app/views/puppetclasses/_form.html.erb:
```diff
@@ -34,7 +34,7 @@
               <span class="input-group-addon">@</span>
               <%= select_tag "environment_filter", options_from_collection_for_select(@puppetclass.environments, "id", "name"),
-                             :include_blank => "All Environments - (Not filtered)",:class=>'form-control', :onchange=>'filterByEnvironment(this)'%>
+                             :prompt => _("All Environments - (Not filtered)"),:class=>'form-control', :onchange=>'filterByEnvironment(this)'%>
```
Keep sentence case here, "All environments (not filtered)"

orrabin commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -7,6 +7,11 @@ $(function() {
+  $('#smart_class_param').parents('form').on('submit', function(){
+    build_match();
+  })
+  $('.matcher_key').select2('destroy');
```
@shlomizadok this is not destroying the select2 and when using select2 and adding to the options I get a list of simple text instead of the option list

orrabin commented:
@domcleal I addressed most of your issues. I'm still having some problems with  the css and I need to find a solution for the location of the validation section.

Trying to add select2 to the matcher created problems since the list of orders is dynamic and adding to the selection list turned it into a list of text instead of a selection list.
The list should be pretty short in most cases and I'm not sure it is worth it to try and fix this for dynamic lists in this PR. We will most likely hit this issue again when making a list of options to match and the solution should fix both lists.

domcleal commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -217,3 +222,60 @@ function validatorTypeSelected(item){
+    matcher_key.empty();
+    matcher_key.append("<option></option>");
+    $.each(order, function (index, value) {
+      matcher_key.append($("<option>", {value: encodeURI(value), html: encodeURI(value)}));
```
Not sure this encoding's right - you want to HTML escape it, not URI encode it.

orrabin replied:
I tried to use `matcher.closest('.matcher-parent').find('#order').html()` but that deleted anything that needed escaping from the order list.
`$(value)` is empty so I can't use `$(value).html()` and `value.escape()` is deprecated: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/escape

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,62 @@
+    </fieldset>
+    </br>
+    <fieldset>
+      <legend><%= _("Specify Matcheres") %>  <%= documentation_button('4.2.6SmartMatchers') %></legend>
```
"Matchers"

domcleal commented:
Okay, no worries about select2.  Leaving as WoC for the work still to do, and a few extra comments above.

waldenraines commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -217,3 +222,60 @@ function validatorTypeSelected(item){
+
+function match_to_key_value(match){
+  var keys = [], values = [],
+    split_matcher = match.replace(/ /g,'').split(/[,=]/);
```
Can you use the `const`s defined above for this `split()`?

orrabin replied:
I'd have to put a space or something between them and that wouldn't work.
Is there another way to put both of them there?

waldenraines replied:
I think you could do something like this (untested):

``` javascript
var regex = new RegExp('[' + KEY_DELM + EQ_DELM + ']');
split_matcher = match.replace(/ /g,'').split(regex);
```

orrabin replied:
thanks, I'll change it

orrabin commented:
@domcleal I fixed most of the issues and responded to the rest.
I moved the validators section up but hid it in an expander (the button doesn't look great yet but I wanted your opinion on it's existence). If you change parameter type it is automatically expanded, otherwise it's closed unless you click on it.

orrabin commented:
[test]

domcleal commented:
Yeah, I think putting it behind a hidden section would be reasonable.  Probably keep the button near the _Parameter type_  dropdown would make the most sense, so you can change one and then the other.  Consider auto-expanding it when editing if the validator's enabled?

orrabin commented:
@domcleal I turned the validator section's title to a dropdown that also opens when parameters type is changed.

waldenraines commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,66 @@
+  </fieldset>
+  <fieldset>
+    <legend data-toggle="collapse" data-target="#optional_input_validators"><%= _("Optional input validator") %>
+      <span class="caret"></span></legend>
```
I would recommend that this legend and the caret use `cursor: pointer;` (perhaps via some already defined class?) as it makes it easier to tell that you can click on this to expand.

waldenraines commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -7,8 +7,21 @@ $(function() {
+function toggleCaret(e) {
+  $(e.target)
+    .parent()
+    .toggleClass('dropup ');
```
As mentioned over IRC, you could remove all of this toggle JS by extending the bootstrap 3 class.  Something like this (untested):

``` scss
@import 'bootstrap';

.collapsed .caret {
  @extend .dropup .caret;
}
```

domcleal commented:
This PR's in an unmergeable state at the moment, please rebase it and check tests pass.

orrabin commented:
@waldenraines with the help of @tbrisker the `caret` is now turning on expand/collapse using css only

domcleal commented:
1) spaces in matcher values seem to be stripped out on page load.  I have a few host groups with spaces in the name, and when viewing the page they simply load as "hostgroup" = "Test2Test3" rather than "Test2 Test3", so when saving, validation fails as the host group can't be found.

2) Inconsistent capitalisation in section headings, prefer sentence case (first upper, rest lower).

![screenshot from 2015-07-20 13 23 10](https://cloud.githubusercontent.com/assets/482089/8775695/816a6b56-2ee2-11e5-9ac0-a92ed97c09f8.png)

3) "Override the Puppet class default value" should probably read "Override the default value of the Puppet class parameter." as it's a per-parameter default.

4) Add full stops/periods to the end of sentences that introduce each section.

5) Is it possible to increase the size of the dropdown arrow next to "Optional input validator"?  I missed it at first.  I don't really have any better suggestions, it's just hard to tell what's happening right now.

6) "Override the Puppet class default value" isn't relevant on the smart variables tab, since there's no default from Puppet, it's only defined in Foreman.

7) When there are no matchers, the table headings look a little lost (as they're unused) and are very close to the "Add Matcher" button.  Consider putting in a blank row, grey text and centred (or similar) saying there are no entries and to press the button below?

![screenshot from 2015-07-20 13 26 43](https://cloud.githubusercontent.com/assets/482089/8775747/00202878-2ee3-11e5-8f96-65b76fbb54f3.png)

orrabin commented:
@domcleal I have addressed all of your comments.
Can you please take another look at the matchers and validator sections?
I made the dropdown arrow bigger and moved it to the right so hopefully it's easier to see what is going on and I hid the matchers table completly when there are no overrides

orrabin commented:
failure looks unrelated so [test] again

orrabin commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -217,3 +226,61 @@ function validatorTypeSelected(item){
+    var matcher_key = matcher.find('.matcher_key');
+    var matcher_value = matcher.find('.matcher_value');
+    var order = matcher.closest('.matcher-parent').find('#order').val().split('\n');
+    matcher_key.empty();
```
@domcleal I removed the empty option

tbrisker commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,66 @@
+  </fieldset>
+  <fieldset>
+    <legend class="expander collapsed" data-toggle="collapse" data-target="#optional_input_validators_<%= f.object.id %>">
+      <span class="caret"></span> <%= _("Optional input validator") %></legend>
```
I think this should be made into a helper, i.e. `collapsing_legend` as this seems like a nice UI widget we can reuse in other places for non-vital fieldsets.

tbrisker commented on app/helpers/layout_helper.rb:
```diff
@@ -66,7 +66,16 @@ def checkbox_f(f, attr, options = {}, checked_value = "1", unchecked_value = "0"
+    end
+  end
+
+  def table_checkbox_f(f, attr, options = {}, checked_value = "1", unchecked_value = "0")
```
DRY?

tbrisker commented on app/assets/stylesheets/puppetclasses.scss:
```diff
@@ -2,5 +2,39 @@
+  margin-right: 10px;
+}
+
+.expander {
```
this should move to application.scss so it is available for the helper on other pages as well

tbrisker commented:
A couple of issues I found:
- when attempting to save after adding a matcher that fails validation (e.g. fqdn that doesn't exist) the matchers are hidden along with the validation error.
- matchers on smart variable edit page don't work

orrabin commented:
@tbrisker thanks for the help, can you please test again?

tbrisker commented:
- Smart variable matcher editing is still not working, matcher value is cleared out before submit and a "can't be blank" message is displayed.
- Adding a smart variable to a puppet class that has a validation error causes the submit to fail but the new variable is not displayed so seeing or fixing the error is not possible.

orrabin commented:
@tbrisker the second issue you mentioned is a bug in develop too, if it's not open I'll open an issue but it should be fixed in another PR. I'll take a look at the first issue.

tbrisker commented:
Changing the order field (editing an existing line) does not clear existing matchers, causing the matchers to be wrong

orrabin commented:
@domcleal I need to readd the empty option because of @tbrisker's last comment - when there is only one possibility in the order and you change it all the matchers will be updated with the only option left since there is no blank option for the matchers dropdown.

tbrisker commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -217,3 +226,61 @@ function validatorTypeSelected(item){
+    var order = matcher.closest('.matcher-parent').find('#order').val().split('\n');
+    matcher_key.empty();
+    $.each(order, function (index, value) {
+      matcher_key.append($("<option>", {value: encodeURI(value), html: encodeURI(value)}));
```
this causes spaces to show in the dropdown as %20

domcleal replied:
Yeah, I flagged this earlier on, I think @orrabin was struggling to find a way to escape the HTML (as if you don't it allows script injection) and URI encoding was the closest?

orrabin commented:
@domcleal sorry, didn't see your comment when I pushed the changes. I was struggling to escape the HTML there, @tbrisker is looking for another solution

domcleal commented on app/views/lookup_keys/_fields.html.erb:
```diff
@@ -1,60 +1,66 @@
+    </br>
+  </fieldset>
+  <fieldset>
+    <%= collapsing_legend _("Optional input validator"), "#optional_input_validators_#{f.object.id}", "collapsed" %>
```
Could you add a little spacing between the collapsing character and the title?  There's only a very narrow gap.  Otherwise looks much better!

![screenshot from 2015-07-28 10 50 30](https://cloud.githubusercontent.com/assets/482089/8928476/a270b302-3516-11e5-83cb-25bc0cc2ab9e.png)

domcleal commented:
1) The tooltip sizes are mostly far too small and some of the text is in bold, though I think this is mostly a pre-existing issue.

![screenshot from 2015-07-28 10 55 07](https://cloud.githubusercontent.com/assets/482089/8928576/3eeeddd0-3517-11e5-883f-0bc36534e502.png)

2) It still seems possible to save the form with a blank value in the matcher/attribute dropdown list.  I saw your earlier comment about not being able to remove, that's fine, but we should probably have some validation in place.

orrabin commented:
@domcleal the sizes are a problem today too and the bold is a result of putting the tooltip in the table header.
I added rejections on matchers with an empty key (empty value was already being rejected).
Also, @tbrisker found a different way to escape html for the priority ordering.

domcleal commented:
Please check the test failure.

orrabin commented:
@domcleal I fixed rubocop errors and rebased

domcleal commented on app/models/lookup_key.rb:
```diff
@@ -7,6 +7,7 @@ class LookupKey < ActiveRecord::Base
 
   KEY_DELM = ","
   EQ_DELM  = "="
+  VALUE_REGEX = Regexp.new("\\A[\\w\\s]+#{EQ_DELM}[\\w\\s]+(#{KEY_DELM}[\\w\\s]+#{EQ_DELM}[\\w\\s]+)*\\Z")
```
I think the `[\w\s]` is too restrictive here, and it's (quietly!) preventing me from saving a matcher of: `hostgroup=Foo/Bar`, since `/` isn't in the word character class.  I'd suggest being a lot more liberal, accepting anything except KEY_DELM perhaps?

domcleal replied:
Second note, you might want to simply use `/\A... \z/` rather than Regexp.new, then you won't have to escape all of those slashes.

domcleal commented on app/models/lookup_key.rb:
```diff
@@ -125,6 +126,12 @@ def path=(v)
   end
 
+  def reject_invalid_lookup_values(attributes)
+    attributes[:match].empty? || attributes[:match].match(VALUE_REGEX).blank? ||
```
Is this covered by a functional or unit test anywhere so we know that ordinary nested attributes will get accepted?  e.g. if the regexp changed to match nothing at all, will we know?

A test for the case you're fixing (where the matcher attribute's nil) might be useful.

domcleal commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -1,13 +1,41 @@
+      </td>
+      <td>
+        <span class="help-block">
+            <%= link_to_remove_fields('', f) %>
```
4-space indentation should be 2-space.  Please just recheck this file, there are a few instances where it's inconsistent.

domcleal commented on app/views/lookup_keys/_values.html.erb:
```diff
@@ -0,0 +1,21 @@
+<tbody>
+
+<%= f.fields_for :lookup_values do |lookup_values| %>
+    <%= render 'lookup_keys/value', :f => lookup_values, :is_param => is_param, :is_template => false %>
```
indentation

domcleal commented on app/assets/javascripts/lookup_keys.js:
```diff
@@ -74,13 +78,15 @@ function add_child_node(item) {
       $('#smart_vars .smart-var-tabs').prepend(pill);
       field = $('#smart_vars .stacked-content').prepend($(content).addClass('active'));
       $('#smart_vars .smart-var-tabs li.active a').show('highlight', 500);
+    } else  if (assoc == 'lookup_values') {
```
two spaces between "else if"

domcleal commented on app/models/lookup_value.rb:
```diff
@@ -13,7 +13,7 @@ class LookupValue < ActiveRecord::Base
   validate :validate_list, :validate_regexp, :ensure_fqdn_exists, :ensure_hostgroup_exists
 
-  attr_accessor :host_or_hostgroup
+  attr_accessor :host_or_hostgroup, :matcher_key, :matcher_value
```
Where are these used?  Commenting them out seems to change nothing, AFAICT.

orrabin commented:
@domcleal I added tests and moved the regex check to `lookup_value` so there will be a validation instead on silently rejecting.

domcleal commented on app/helpers/layout_helper.rb:
```diff
@@ -193,28 +193,61 @@ def autocomplete_f(f, attr, options = {})
+    size_class  = options.delete(:size) || "col-md-4"
+    label = add_label(options, f, attr)
+
+    if !table_field
```
Nitpick, use `unless` or preferably swap the two blocks of code around so this reads `if table_field`.  (There's a rubocop TODO to enforce this.)

domcleal commented:
Thanks, that works much better.

waldenraines commented on app/assets/javascripts/jquery.extentions.js:
```diff
@@ -22,3 +22,7 @@ jQuery.humanize = function(str) {
+
+jQuery.escapeHTML = function(str) {
+  return $("<div></div>").text(str).html();
+};
```
We could consider using [`_.escape()`](http://underscorejs.org/#escape) if we wanted to bring in underscore.js (really useful for other utility type things as well) so we don't have to maintain this.

At the very least we should use one of the suggestions [here](http://stackoverflow.com/a/6020820).  Note the first comment on that answer.

orrabin commented:
@domcleal I switched between the two blocks
@waldenraines I added `underscore.js` as you requested, can you please test the escaping? it looks the same as before to me.

domcleal commented:
> I added underscore.js

Could you use underscore-rails please?  Add it to bundler.d/assets.rb with suitable pinning.

unorthodoxgeek commented:
@domcleal - I strongly disagree with adding underscore-rails. Gems which don't do anything except add an asset add complexity, consume memory and make load times painful (once they add up). There's absolutely no need to add a ruby gem just to add an asset.

domcleal commented:
This is how we currently manage as many assets as possible in Foreman and not about to debate the advantages/disadvantages of it in this PR - the PR should use our current style.  If you'd like to change this for some reason, bring it up on foreman-dev please.

orrabin commented on bundler.d/assets.rb:
```diff
@@ -18,4 +18,5 @@
   gem 'jquery_pwstrength_bootstrap', '~> 1.2'
   gem 'jquery-turbolinks', '~> 2.1'
   gem 'select2-rails', '~> 3.5'
+  gem 'underscore-rails', '~> 1.8'
```
@domcleal latest version is 1.8.3

waldenraines commented:
> @waldenraines I added underscore.js as you requested, can you please test the escaping? it looks the same as before to me.

This is working for me as well, thanks for making the change.

domcleal commented on app/assets/stylesheets/puppetclasses.scss:
```diff
@@ -2,5 +2,26 @@
+  }
+}
+
+.form-group button[data-original-title="Full screen"]{
```
This won't work in anything other than English, I'd suggest replacing it with a class.

domcleal commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -1,13 +1,41 @@
+  </tr>
+<% end %>
+<% if is_template %>
+    </table>
```
indentation's still off here

domcleal commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -1,13 +1,41 @@
+  </td>
+  <td>
+    <span class="help-block">
+        <%= link_to_remove_fields('', f) %>
```
indentation's still off here

domcleal commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -1,13 +1,41 @@
+  <tr class="has-error">
+    <td colspan="5">
+      <span class="help-block">
+          <%= f.object.errors.full_messages.to_sentence %>
```
indentation's still off here

domcleal commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -1,13 +1,41 @@
-                    :onchange=>'toggleUsePuppetDefaultValue(this, "value")') if is_param %>
-  <%= textarea_f f, :value, :rows => :auto, :value => f.object.value_before_type_cast, :size => "col-md-8",
+<% if is_template %>
+    <table>
```
indentation's still off here

domcleal commented on app/views/lookup_keys/_value.html.erb:
```diff
@@ -1,13 +1,41 @@
+  </td>
+  <td class="ca">
+    <%= f.check_box :use_puppet_default, :'data-property' => 'use_puppet_default',
+                    :disabled => (not is_param),
```
Nitpick: prefer `!` as a logic instead of flow control operator

orrabin replied:
I should have added `if is_param` to display this only on the class parameters tab and not just disabled the checkbox, I'll fix it.

domcleal commented on test/unit/lookup_value_test.rb:
```diff
@@ -154,13 +154,31 @@ def valid_attrs2
     refute value.valid?
   end
 
+  test "lookup value should allow valid key" do
```
Could you also add a test please for multi-value lookup values?  (hostgroup=Common,domain=foo)

orrabin commented:
@domcleal thanks for the comments, I fixed indentation, removed use_puppet_default from smart variables completely, added 2 tests for valid and invalid multiple keys and rebased.

domcleal commented on app/assets/stylesheets/puppetclasses.scss:
```diff
@@ -2,5 +2,26 @@
+  }
+}
+
+.btn-fullscreen{
```
I think this should be in application as it affects the whole app.  You could perhaps restrict it to lookup-keys-container, but it seems to also be necessary on the Parameters tab of host#edit, though I can't tell why it's now necessary to make that look the same as it did.

orrabin replied:
Agreed, I'll move this as I need it for both places you mentioned and it doesn't affect provision templates since there is no button close to the full screen one on the right side.

orrabin commented:
@domcleal I moved the css changes to application.

domcleal commented:
Merged as 09ce8a6, thanks @orrabin.
